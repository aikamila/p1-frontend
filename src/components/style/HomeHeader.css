.home-header__background {
  border-bottom: 1px solid lightgray;
  height: 54px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.home-header__background > div {
  display: inline;
}
.home-header__nav {
  display: inline-flex;
  align-items: center;
}

@media screen and (min-width: 481px) {
  .home-header__img1,
  .home-header__img2 {
    height: 47px;
    margin-left: 4px;
  }
}

@media screen and (max-width: 480px) {
  .home-header__img1,
  .home-header__img2 {
    height: 37px;
    margin-left: 4px;
  }
}

.home-header__nav > a {
  text-decoration: none;
  color: var(--nice-blue);
}

.home-header__nav > a {
  font-size: 1.5em;
}

@media screen and (min-width: 481px) {
  .home-header__nav > a,
  .home-header__nav > button {
    margin-left: 12px;
  }
  .home-header__nav > button {
    margin-right: 20px;
  }
}

@media screen and (max-width: 480px) {
  .home-header__nav > a,
  .home-header__nav > button {
    margin-left: 7px;
  }
  .home-header__nav > button {
    margin-right: 5px;
  }
}

.home-header__nav > button {
  border-radius: 3px;
  border-color: transparent;
  background-color: var(--nice-blue);
  color: white;
  font-weight: 400;
}

.home-header__nav > button:hover {
  cursor: pointer;
}
