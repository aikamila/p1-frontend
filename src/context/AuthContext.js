import { createContext, useState, useEffect, useCallback } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { getItemAndCheckExpiry, setItemwith8MonthExpiry } from "../utils/LocalStorage";
import { calculateTime } from "../utils/Timing";
import { decodeUser } from "../utils/JWTDecode";

const AuthContext = createContext()

export default AuthContext;

export const AuthProvider = ({children}) => {

    const history = useHistory()

    let location = useLocation()

    const [authTokens, setAuthTokens] = useState(null)
    const [userId, setUserId] = useState(null)

    const [loginEmail, setLoginEmail] = useState('')
    const [loginPassword, setLoginPassword] = useState('')

    // const [pwResetEmail, setPwResetEmail] = useState('')
    // const [pwResetPassword, setPwResetPassword] = useState('')
    // const [pwResetPasswordSecond, setPwResetPasswordSecond] = useState('')

    const [loginError, setLoginError] = useState(null) // invalid credentials
 
    const [authServerError, setAuthServerError] = useState(null) // server error while logging in
    const [authSignUpServerError, setAuthSignUpServerError] = useState(null) // server error while signing up
    const [unexpectedLogoutError, setUnexpectedLogoutError] = useState('Something went wrong and you were logged out.')
    const [initialVerificationError, setInitialVerificationError] = useState(null)
    const [initialVerificationServerError, setInitialVerificationServerError] = useState(null)
    const [initialVerificationSuccessful, setInitialVerificationSuccessful] = useState(true)

    const [loginButtonDisabled, setLoginButtonDisabled] = useState(false)

    const BACKEND_DOMAIN = 'https://arcane-spire-03245.herokuapp.com/'
    const TOKENS_OBTAIN_URL = BACKEND_DOMAIN + 'api/user/token/'
    const TOKENS_REFRESH_URL = BACKEND_DOMAIN + 'api/user/token/refresh/'
    const TOKENS_BLACKLIST_URL = BACKEND_DOMAIN + 'api/user/token/blacklist/'
    const INITIAL_EMAIL_VERIFICATION = BACKEND_DOMAIN + 'api/user/token/verification/'

    const login = useCallback(async (e) => {
        e.preventDefault();
        setLoginButtonDisabled(true)
        try{
            const response = await fetch(TOKENS_OBTAIN_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'email': loginEmail, 'password': loginPassword})
            })
            const data = await response.json()
            if (response.status === 401) {
                throw Error('Invalid credentials')
            }else if (response.status === 200){
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setLoginPassword('')
                history.push('/home')
            }else if(loginEmail.length === 0 || loginPassword.length === 0){
                throw Error('Invalid credentials')
            }else{
                throw Error('Something else happened')
            }
        }catch (err){
            switch (err.message){
                case 'Invalid credentials':
                    setLoginError(err.message)
                    setLoginPassword('')
                    break;
                default:
                    setAuthServerError('Upps... unexpected problem with the server occured. Try to refresh the site and log in again :)')
            }
        }
        setLoginButtonDisabled(false)
    }, 
    [TOKENS_OBTAIN_URL, loginEmail, loginPassword, history]
    )

    const logout = useCallback(async () => {
        try{
            await fetch(TOKENS_BLACKLIST_URL,{
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'refresh': authTokens.refresh})
            })}catch(err){
                // doing nothing if the blacklisting of tokens goes wrong on the backend
            }
        const tokens = localStorage.getItem('authTokens')
        if (tokens){
            localStorage.removeItem('authTokens')
        }
        setAuthTokens(null)
        setUserId(null)
        history.push('/')
    }, [TOKENS_BLACKLIST_URL])

    const updateToken = useCallback(async () => {
        try{
            const tokens = getItemAndCheckExpiry('authTokens')
            // we get them from the local storage bc when the user comes back, they don't have the hook set (so we would send null :( )
            const response = await fetch(TOKENS_REFRESH_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'refresh': tokens.refresh})
            })
            const data = await response.json()
            if (response.status !== 200) {
                throw Error('Invalid credentials')
            }else{
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setUserId(decodeUser(data))
            }
        }catch(err){
            logout(); 
            setUnexpectedLogoutError('Something went wrong and you were logged out.')
            setTimeout(() => {
                setUnexpectedLogoutError(null)
            }, 5000)
        }
    }, [TOKENS_REFRESH_URL, logout]
    )

    const verifyEmail = async (id, token) => {
        try{
            const response = await fetch(INITIAL_EMAIL_VERIFICATION, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'id': id, 'token': token})
                })
            if(response.status === 401){
                throw Error('Invalid link')
            }else if(response.status !== 200){
                throw Error("Something else happened")
            }
            const data = await response.json()
            if(response.status===200){
                await logout(); // if there is a user that is logged in, their token should be blacklisted
                setAuthTokens(data);
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setInitialVerificationSuccessful(true)
                setTimeout(() => {
                    setInitialVerificationSuccessful(false)
                }, 5000)
                history.push('/home')
            }
            }catch(err){
                if(err.message === 'Invalid link'){
                    setInitialVerificationError('The link is invalid. If you have already used it, just log into your account.')
                }else(
                    setInitialVerificationServerError('Something went wrong. Please use the link once again.')
                )
                history.push("/verification/unsuccessful")
            }
        
    }

    useEffect(() => {
        const fourMinutes = calculateTime(0, 4, 0)
        let interval = setInterval(() => {
            if (authTokens){
                updateToken()
            }
        }, fourMinutes)
        return () => clearInterval(interval)
    }, [authTokens, updateToken])

    useEffect(() => {
        const updateTokenAtTheBeginning = async () => {
            const tokens = getItemAndCheckExpiry('authTokens')
            if(tokens !== null){
                await updateToken();
            }
        }   
        updateTokenAtTheBeginning()
    }, [updateToken])

    useEffect(() => {
        setLoginPassword('')
        setLoginError(null)
        setAuthSignUpServerError(null)
        setAuthServerError(null)
    }, [location])

    let contextData = {
        login: login,
        loginEmail: loginEmail,
        setLoginEmail: setLoginEmail,
        loginPassword: loginPassword,
        setLoginPassword: setLoginPassword,
        loginError: loginError,
        authTokens: authTokens, 
        logout: logout,
        authServerError: authServerError,
        unexpectedLogoutError: unexpectedLogoutError,
        verifyEmail: verifyEmail,
        initialVerificationError: initialVerificationError,
        initialVerificationServerError: initialVerificationServerError,
        initialVerificationSuccessful: initialVerificationSuccessful,
        setAuthSignUpServerError: setAuthSignUpServerError,
        authSignUpServerError: authSignUpServerError,
        BACKEND_DOMAIN: BACKEND_DOMAIN,
        setLoginError: setLoginError,
        userId: userId,
        updateToken: updateToken,
        loginButtonDisabled: loginButtonDisabled,
    }

    return(
        <AuthContext.Provider value={contextData}>
            {children}
        </AuthContext.Provider>
    )}