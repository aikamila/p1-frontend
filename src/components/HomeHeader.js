import React, {useContext} from 'react'
import AuthContext from '../context/AuthContext'
import { Link } from 'react-router-dom'

const HomeHeader = () => {
    let { logout, userId} = useContext(AuthContext)
    return (
        <header>
            <nav>
                <Link to="/home/post/add">Add a post</Link>
                <Link to="/home">Homepage</Link>
                <Link to={`/home/account/${userId}`}>My account</Link>
                <button onClick={logout}>Log out</button>
            </nav>
        </header>
    )
}

export default HomeHeader
