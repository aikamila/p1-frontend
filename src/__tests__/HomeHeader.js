import { render, screen } from '@testing-library/react'
import react from 'react'
import HomeHeader from '../components/HomeHeader'
import AuthContext from '../context/AuthContext'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'

const renderHomeHeader = (disabled) => {
    const history = createMemoryHistory()
    return render(
        <Router history={history}>
            <AuthContext.Provider value={{logout: jest.fn(), userId: 12}}>
                <HomeHeader disabled={disabled}>
                </HomeHeader>
            </AuthContext.Provider>
        </Router>
    )
}


test("all elements of the header are fully accessible", () => {
    renderHomeHeader();
    const logoutButton = screen.getByRole("button", {name: /log out/i})
    expect(logoutButton).toBeInTheDocument()
    expect(logoutButton).not.toBeDisabled()
    const linkAdd = screen.getByRole("link", {name: /add a post/i})
    const linkAccount = screen.getByRole("link", {name: /my account/})
    const linkHomepage = screen.getByRole("link", {name: /homepage/i})
    expect(linkAdd).toBeInTheDocument()
    expect(linkAdd).toHaveAttribute('href', '/home/post/add')
    expect(linkAdd).toHaveStyle('pointerEvents: auto')
    expect(linkAccount).toBeInTheDocument()
    expect(linkAccount).toHaveAttribute('href', '/home/account/12')
    expect(linkAccount).toHaveStyle('pointerEvents: auto')
    expect(linkHomepage).toBeInTheDocument()
    expect(linkHomepage).toHaveStyle('pointerEvents: auto')
    expect(linkHomepage).toHaveAttribute('href', '/home')
    expect(screen.getByTitle(/my account/i)).toBeInTheDocument()
    expect(screen.getByTitle(/add a post/i)).toBeInTheDocument()
    expect(screen.getByTitle(/homepage/i)).toBeInTheDocument()
    expect(screen.getByRole("img", {name:/shangify brand logo/i})).toBeInTheDocument()
})

test("all controls are disabled", () => {
    renderHomeHeader(true);
    const logoutButton = screen.getByRole("button", {name: /log out/i})
    expect(logoutButton).toBeDisabled()
    const linkAdd = screen.getByRole("link", {name: /add a post/i})
    const linkAccount = screen.getByRole("link", {name: /my account/})
    const linkHomepage = screen.getByRole("link", {name: /homepage/i})
    expect(linkAdd).toHaveStyle('pointerEvents: none')
    expect(linkAccount).toHaveStyle('pointerEvents: none')
    expect(linkHomepage).toHaveStyle('pointerEvents: none')
})
