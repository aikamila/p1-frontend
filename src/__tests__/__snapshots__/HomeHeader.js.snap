// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all elements are rendered correctly 1`] = `
<header
  className="home-header__background"
  role="banner"
>
  <div>
    <img
      alt="shangify brand logo"
      className="logo__img1"
      src="shangify_logo.png"
    />
    <img
      alt=""
      className="logo__img2"
      src="shangify_brand_name.png"
    />
  </div>
  <nav
    aria-label="main navigation"
    className="home-header__nav"
    role="navigation"
  >
    <a
      aria-label="add a post"
      href="/home/post/add"
      onClick={[Function]}
      style={
        Object {
          "pointerEvents": "auto",
        }
      }
      title="Add a post"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4z"
        />
        <path
          d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm0 18c-4.411 0-8-3.589-8-8s3.589-8 8-8 8 3.589 8 8-3.589 8-8 8z"
        />
      </svg>
    </a>
    <a
      aria-label="homepage"
      href="/home"
      onClick={[Function]}
      style={
        Object {
          "pointerEvents": "auto",
        }
      }
      title="Homepage"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 22h14a2 2 0 0 0 2-2v-9a1 1 0 0 0-.29-.71l-8-8a1 1 0 0 0-1.41 0l-8 8A1 1 0 0 0 3 11v9a2 2 0 0 0 2 2zm5-2v-5h4v5zm-5-8.59 7-7 7 7V20h-3v-5a2 2 0 0 0-2-2h-4a2 2 0 0 0-2 2v5H5z"
        />
      </svg>
    </a>
    <a
      aria-label="my account"
      href="/home/account/12"
      onClick={[Function]}
      style={
        Object {
          "pointerEvents": "auto",
        }
      }
      title="My account"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 14s-1 0-1-1 1-4 6-4 6 3 6 4-1 1-1 1H3zm5-6a3 3 0 1 0 0-6 3 3 0 0 0 0 6z"
        />
      </svg>
    </a>
    <button
      disabled={false}
      onClick={[MockFunction]}
    >
      Log out
    </button>
  </nav>
</header>
`;
